File: 0002/0002_add_two_numbers.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var result []int

	for l1 != nil && l2 != nil {
		result = append(result, l1.Val+l2.Val)
		l1 = l1.Next
		l2 = l2.Next
	}

	for l1 != nil {
		result = append(result, l1.Val)
		l1 = l1.Next
	}

	for l2 != nil {
		result = append(result, l2.Val)
		l2 = l2.Next
	}

	// build singly linked list
	var lh *ListNode = nil // head
	var lt *ListNode = nil // tail

	for i := 0; i < len(result); i++ {
		if result[i] >= 10 {
			if i == len(result)-1 {
				result = append(result, 0)
			}
			result[i+1]++
			result[i] %= 10
		}

		if lh == nil {
			lh = &ListNode{
				Val:  result[i],
				Next: nil,
			}
			lt = lh
		} else {
			node := &ListNode{
				Val:  result[i],
				Next: nil,
			}
			lt.Next = node
			lt = node
		}
	}

	return lh
}

func build_list(nums []int) *ListNode {
	if len(nums) == 0 {
		return nil
	}

	var l *ListNode = &ListNode{
		Val:  nums[0],
		Next: nil,
	}

	p := l
	for i := 1; i < len(nums); i++ {
		n := &ListNode{
			Val:  nums[i],
			Next: nil,
		}
		p.Next = n
		p = n
	}

	return l
}

func print_list(l *ListNode) {
	for l != nil {
		fmt.Print(l.Val)
		l = l.Next
	}
	fmt.Print("\n")
}

func main() {
	// Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
	// Output: 7 -> 0 -> 8
	print_list(addTwoNumbers(
		build_list([]int{2, 4, 3}),
		build_list([]int{5, 6, 4}),
	))
}

File: 0002/0002_add_two_numbers.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode {
            next: None,
            val
        }
    }
}

struct Solution;

impl Solution {
    pub fn add_two_numbers(
        list1: Option<Box<ListNode>>,
        list2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let mut dummy_head = ListNode::new(0);
        let mut current = &mut dummy_head;
        let mut p = list1;
        let mut q = list2;
        let mut carry: i32 = 0;

        while p != None || q != None {
            let sum = match (&p, &q) {
                (Some(list1), Some(list2)) => list1.val + list2.val + carry,
                (Some(list1), None) => list1.val + carry,
                (None, Some(list2)) => list2.val + carry,
                (None, None) => carry,
            };
            carry = sum / 10;
            current.next = Some(Box::new(ListNode::new(sum % 10)));
            current = current.next.as_mut().unwrap();
            p = if p != None { p.unwrap().next } else { p };
            q = if q != None { q.unwrap().next } else { q };
        }
        if carry > 0 {
            current.next = Some(Box::new(ListNode::new(carry)))
        }
        dummy_head.next
    }
}

fn main() {
    Solution::add_two_numbers(None, None);
}
